### PR TITLE
Fix deadlock in client.next_update (#83)

### LINKED ISSUE #83: Deadlock of chat_hashes between next_update and process_socket_updates
Resource `chat_hashes` is locked first by `client.next_update`
https://github.com/Lonami/grammers/blob/91660ceace62d86b84b05268b91fc84279883cb1/lib/grammers-client/src/client/updates.rs#L67-L70

and then by `client.process_socket_updates`
https://github.com/Lonami/grammers/blob/91660ceace62d86b84b05268b91fc84279883cb1/lib/grammers-client/src/client/updates.rs#L109

invoked by the `await` call on `self.invoke` (I believe)
https://github.com/Lonami/grammers/blob/91660ceace62d86b84b05268b91fc84279883cb1/lib/grammers-client/src/client/updates.rs#L74

<details>
<summary>Logs</summary>

```
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.updates for client.next_update
TRACE [grammers/lib/grammers-client/src/utils.rs:137] unlocking client.updates for client.next_update
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.message_box for client.next_update
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.chat_hashes for client.next_update/get_channel_difference
TRACE [grammers/lib/grammers-client/src/utils.rs:137] unlocking client.message_box for client.next_update
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.request_tx for invoke
DEBUG [grammers/lib/grammers-mtsender/src/lib.rs:113] enqueueing request updates.getChannelDifference to be serialized
TRACE [grammers/lib/grammers-client/src/utils.rs:137] unlocking client.request_tx for invoke
TRACE [grammers/lib/grammers-client/src/utils.rs:165] try-async-locking client.sender for client.step (success)
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:225] reading bytes from the network
TRACE [grammers/lib/grammers-client/src/utils.rs:204] async-unlocking client.sender for client.step
TRACE [grammers/lib/grammers-client/src/utils.rs:165] try-async-locking client.sender for client.step (success)
DEBUG [grammers/lib/grammers-mtsender/src/lib.rs:312] serialized request 3173d78 (updates.getChannelDifference) with MsgId(-------------------)
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:240] reading bytes and sending up to 196 bytes via network
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:364] written 196 bytes to the network (196/196)
DEBUG [grammers/lib/grammers-mtsender/src/lib.rs:381] sent request with MsgId(-------------------)
TRACE [grammers/lib/grammers-client/src/utils.rs:204] async-unlocking client.sender for client.step
TRACE [grammers/lib/grammers-client/src/utils.rs:165] try-async-locking client.sender for client.step (success)
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:225] reading bytes from the network
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:333] read 628 bytes from the network
TRACE [grammers/lib/grammers-mtsender/src/lib.rs:335] trying to unpack buffer of 628 bytes...
DEBUG [grammers/lib/grammers-mtsender/src/lib.rs:409] deserializing valid transport packet...
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.message_box for client.process_socket_updates
TRACE [grammers/lib/grammers-client/src/utils.rs:106] locking client.chat_hashes for client.process_socket_updates
```

</details>

### PR DESCRIPTION
**Pull request by novelhawk**
I've used code blocks this time to prevent more errors from happening with hard to follow drop-calls.
Not sure if it adheres to the repo's code-style guidelines but I believe it's more readable. LMK if you prefer a normal if let syntax with drops after the blocks.
(The drop order should work as expected. See [playground test](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dfb51fca618badfcf7eb063b299f4eb3))

Fixes #83

### CODE REVIEW

File: lib/grammers-client/src/client/updates.rs [hunk 1]
```diff
@@ -64,27 +64,30 @@ impl Client {
...
-                    message_box.apply_channel_difference(request, response, &mut chat_hashes);
+                    let mut chat_hashes = self
+                        .0
+                        .chat_hashes
+                        .lock("client.next_update/get_channel_difference");
+
+                    message_box.apply_channel_difference(request, response, &mut chat_hashes)
+                };
 
                 self.extend_update_queue(updates, ChatMap::new(users, chats));
```
Comment (novelhawk) [inline lib/grammers-client/src/client/updates.rs:88]: I'm dropping message_box and chat_hashes before this call since neither are passed as arguments.
I'm assuming this is fine but let me know if either are required to live through `extend_update_queue`

Comment (Lonami) [reply lib/grammers-client/src/client/updates.rs:88]: Nah as long as it compiles it's good.
